dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.42/V2_42_26__Add_not_null_constraint_to_trackedentity.sql

```diff
@@ -0,0 +1,35 @@
+DO $$
+DECLARE
+inconsistent_records_count INTEGER;
+BEGIN
+
+SELECT COUNT(*) INTO inconsistent_records_count FROM trackedentity WHERE trackedentitytypeid IS NULL;
+
+IF inconsistent_records_count > 0 THEN
+        RAISE NOTICE 'Inconsistencies found: %', inconsistent_records_count;
+
+    UPDATE trackedentity
+    SET trackedentitytypeid = (
+        SELECT program.trackedentitytypeid
+        FROM program
+                 JOIN enrollment ON program.programid = enrollment.programid
+        WHERE enrollment.trackedentityid = trackedentity.trackedentityid
+    )
+    WHERE trackedentitytypeid IS NULL
+      AND EXISTS (
+            SELECT 1
+            FROM enrollment
+                     JOIN program ON enrollment.programid = program.programid
+            WHERE enrollment.trackedentityid = trackedentity.trackedentityid
+        );
+
+    ALTER TABLE IF EXISTS trackedentity ALTER COLUMN trackedentitytypeid SET NOT NULL;
+ELSE
+        RAISE NOTICE 'No inconsistencies found, trackedentitytypeid is already populated.';
+        ALTER TABLE IF EXISTS trackedentity ALTER COLUMN trackedentitytypeid SET NOT NULL;
```

Comment:
I did not understand why do we have two ALTER COLUMN SET NOT NULL on the same table-column combination?

Reply:
If the initial migration fails due to invalid data, it will raise an exception, prompting the user to correct or remove the problematic data. Once the issues are resolved, the same migration can be reapplied, and the ELSE condition will enforce the NOT NULL constraint on the column.

Reply:
Then, can we move the ALTER TABLE to outside the IF ELSE condition, since it is executed at the end anyway?